File: member-service/src/main/java/com/hedvig/memberservice/services/BankIdService.java
package com.hedvig.memberservice.services;

import static com.hedvig.memberservice.query.CollectType.RequestType;

import com.hedvig.external.bankID.bankidTypes.CollectResponse;
import com.hedvig.external.bankID.bankidTypes.OrderResponse;
import com.hedvig.memberservice.query.CollectRepository;
import com.hedvig.memberservice.query.CollectType;
import com.hedvig.memberservice.services.bankid.BankIdSOAPApi;
import java.io.UnsupportedEncodingException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class BankIdService {

  private final BankIdSOAPApi bankIdSOAPApi;
  private final CollectRepository collectRepository;
  private final Logger log = LoggerFactory.getLogger(BankIdService.class);

  public BankIdService(BankIdSOAPApi bankIdSOAPApi, CollectRepository collectRepository) {
    this.bankIdSOAPApi = bankIdSOAPApi;
    this.collectRepository = collectRepository;
  }

  public OrderResponse auth(Long memberId) {
    OrderResponse status = bankIdSOAPApi.auth();
    log.info(
        "Started bankId AUTH autostart:{}, reference:{}",
        status.getAutoStartToken(),
        status.getOrderRef());
    trackAuthToken(status.getOrderRef(), memberId);
    return status;
  }

  public OrderResponse sign(String ssn, String userMessage, Long memberId)
      throws UnsupportedEncodingException {
    OrderResponse status = bankIdSOAPApi.sign(ssn, userMessage);
    trackSignToken(status.getOrderRef(), memberId);
    return status;
  }

  private void trackSignToken(String referenceToken, Long memberId) {
    trackReferenceToken(referenceToken, RequestType.SIGN, memberId);
  }

  private void trackAuthToken(String referenceToken, Long memberId) {
    trackReferenceToken(referenceToken, RequestType.AUTH, memberId);
  }

  private void trackReferenceToken(String referenceToken, RequestType sign, Long memberId) {
    CollectType ct = new CollectType();
    ct.token = referenceToken;
    ct.type = sign;
    ct.memberId = memberId;
    collectRepository.save(ct);
  }

  public CollectResponse authCollect(String referenceToken) {
    return bankIdSOAPApi.authCollect(referenceToken);
  }

  public CollectResponse signCollect(String referenceToken) {
    return bankIdSOAPApi.signCollect(referenceToken);
  }
}

File: member-service/src/main/java/com/hedvig/memberservice/services/BankIdService.kt
package com.hedvig.memberservice.services

import com.hedvig.external.bankID.bankId.BankIdApi
import com.hedvig.external.bankID.bankIdTypes.CollectRequest
import com.hedvig.external.bankID.bankIdTypes.OrderAuthRequest
import com.hedvig.memberservice.query.CollectType.RequestType

import com.hedvig.external.bankID.bankIdTypes.CollectResponse
import com.hedvig.external.bankID.bankIdTypes.OrderResponse
import com.hedvig.memberservice.query.CollectRepository
import com.hedvig.memberservice.query.CollectType
import com.hedvig.memberservice.util.logger
import java.io.UnsupportedEncodingException
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service

@Service
class BankIdService(
    private val bankIdApi: BankIdApi,
    private val collectRepository: CollectRepository
) {

    fun auth(memberId: Long?, endUserIp: String?): OrderResponse {
        val status = bankIdApi.auth(OrderAuthRequest(endUserIp))
        logger.debug(
            "Started bankId AUTH autostart:{}, reference:{}",
            status.autoStartToken,
            status.orderRef)
        trackAuthToken(status.orderRef, memberId)
        return status
    }

    private fun trackAuthToken(referenceToken: String, memberId: Long?) {
        trackReferenceToken(referenceToken, RequestType.AUTH, memberId)
    }

    private fun trackReferenceToken(referenceToken: String, sign: RequestType, memberId: Long?) {
        val ct = CollectType()
        ct.token = referenceToken
        ct.type = sign
        ct.memberId = memberId
        collectRepository.save(ct)
    }

    fun authCollect(referenceToken: String): CollectResponse {
        return bankIdApi.collect(CollectRequest(referenceToken))
    }

    fun signCollect(referenceToken: String): CollectResponse {
        return bankIdApi.collect(CollectRequest(referenceToken))
    }

}
